rholang/src/main/scala/coop/rchain/rholang/interpreter/PrettyPrinter.scala

```diff
@@ -0,0 +1,205 @@
+import coop.rchain.models.Channel.ChannelInstance.{ChanVar, Quote}
+import coop.rchain.models.Expr.ExprInstance
+import coop.rchain.models.Expr.ExprInstance._
+import coop.rchain.models.Var.VarInstance
+import coop.rchain.models.Var.VarInstance.{BoundVar, FreeVar, Wildcard}
+import coop.rchain.models._
+import coop.rchain.rholang.interpreter.implicits.ChannelLocallyFree._
+
+object PrettyPrinter {
+  def apply(): PrettyPrinter = PrettyPrinter(0, 0)
+}
+
+case class PrettyPrinter(freeShift: Int, boundShift: Int) {
+
+  val MAX_NEW_VAR_COUNT = 128
+
+  def buildString(e: Expr): String =
+    e.exprInstance match {
+      case ENegBody(ENeg(p)) => "-" + buildString(p.get)
+      case ENotBody(ENot(p)) => "~" + buildString(p.get)
+      case EMultBody(EMult(p1, p2)) =>
+        buildString(p1.get) + " * " + buildString(p2.get)
+      case EDivBody(EDiv(p1, p2)) =>
+        buildString(p1.get) + " / " + buildString(p2.get)
+      case EPlusBody(EPlus(p1, p2)) =>
+        buildString(p1.get) + " + " + buildString(p2.get)
+      case EMinusBody(EMinus(p1, p2)) =>
+        buildString(p1.get) + " - " + buildString(p2.get)
+      case EAndBody(EAnd(p1, p2)) =>
+        buildString(p1.get) + " && " + buildString(p2.get)
+      case EOrBody(EOr(p1, p2)) =>
+        buildString(p1.get) + " || " + buildString(p2.get)
+      case EEqBody(EEq(p1, p2)) =>
+        buildString(p1.get) + " == " + buildString(p2.get)
+      case ENeqBody(ENeq(p1, p2)) =>
+        buildString(p1.get) + " != " + buildString(p2.get)
+      case EGtBody(EGt(p1, p2)) =>
+        buildString(p1.get) + " > " + buildString(p2.get)
+      case EGteBody(EGte(p1, p2)) =>
+        buildString(p1.get) + " >= " + buildString(p2.get)
+      case ELtBody(ELt(p1, p2)) =>
+        buildString(p1.get) + " < " + buildString(p2.get)
+      case ELteBody(ELte(p1, p2)) =>
+        buildString(p1.get) + " <= " + buildString(p2.get)
+      case EListBody(EList(s, _, _, _)) =>
+        "[" + buildSeq(s) + "]"
+      case ETupleBody(ETuple(s, _, _, _)) =>
+        "[" + buildSeq(s) + "]"
+      case ESetBody(ESet(s, _, _, _)) =>
+        "(" + buildSeq(s) + ")"
+      case EMapBody(EMap(kvs, _, _, _)) =>
+        "{ " + ("" /: kvs.zipWithIndex) {
+          case (string, (kv, i)) =>
+            string + buildString(kv.key.get) + ": " + buildString(kv.value.get) + {
+              if (i != kvs.length - 1) ", "
+              else ""
+            }
+        } + " }"
+
+      case EVarBody(EVar(v)) => buildString(v.get)
+      case GBool(b) => b.toString
+      case GInt(i) => i.toString
+      case GString(s) => "\"" + s + "\""
+      case GUri(u) => s"`$u`"
+      // TODO: Figure out if we can prevent ScalaPB from generating
+      case ExprInstance.Empty => "Nil"
+      case _ => throw new Error("Attempted to print unknown Expr type")
+    }
+
+  def buildString(v: Var): String =
+    v.varInstance match {
+      case FreeVar(level) => s"x${freeShift + level}"
```

Comment:
Until we use different variables names as we step inside patterns, these should print as `z`. I can come up with counter examples to where no matter what we pass as freeShift and boundShift, there's a conflict.

Reply:
Won't there be a disconnect between the identifier used for the binding occurrence variable and a bound occurrence of that same variable?

What I imagine happening is:

`for( z0 <- x0 ){ *x0 }` ==> `for ( z1 <- x0 ){ *x1 }`

Reply:
Yes, there will be, but it's better than the alternative.

```for ({for (z0 <-_) { *z0 }} <- x0) { P }```
prints as
```for ({for (x0 <-_) { *x0 }} <- x0 { P }```
but so does:
```for ({for (z0 <-_) { *x0 }} <- x0) { P }```
Like I said, it's only __until__ we use a different variable for each pattern level.
Until then, using z for free and x for bound is at least not ambiguous, even if we can't quite compile the output, at least we can read it unambiguously.

Reply:
I see. Will make changes and start Rot 23.